### PR TITLE
Error handling cleanup in solvers/flattening 7

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

File: src/solvers/flattening/flatten_byte_operators.cpp [hunk 1]
```diff
@@ -199,8 +202,6 @@ exprt flatten_byte_extract(
   // big-endian: (short)concatenation(01,02)=0x0102
   // little-endian: (short)concatenation(03,04)=0x0304
 
-  assert(src.operands().size()==2);
```
Comment (peterschrammel) [inline src/solvers/flattening/flatten_byte_operators.cpp:202]: which check covers this?

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/flattening/equality.cpp [hunk 2]
```diff
@@ -24,10 +24,9 @@ literalt equalityt::equality(const exprt &e1, const exprt &e2)
 
 literalt equalityt::equality2(const exprt &e1, const exprt &e2)
 {
-  const typet &type=e1.type();
+  PRECONDITION(e2.type() != e1.type());
```
Comment (tautschnig) [inline src/solvers/flattening/equality.cpp]: Those better be equal ;-)

File: src/solvers/flattening/equality.cpp [hunk 3]
```diff
@@ -36,9 +35,8 @@ literalt equalityt::equality2(const exprt &e1, const exprt &e2)
 
   // check for boolean equality
 
-  if(type.id()==ID_bool)
-    throw "equalityt got boolean equality";
-  // return lequal(convert(e1), convert(e2));
+  INVARIANT(
+    type.id() != ID_bool, "method shall not be used to compare boolean types");
```
Comment (tautschnig) [inline src/solvers/flattening/equality.cpp]: s/boolean/Boolean/

File: src/solvers/flattening/flatten_byte_operators.cpp (hunk 1 above)
Comment (tautschnig) [reply src/solvers/flattening/flatten_byte_operators.cpp:202]: The fact that the parameter type is `byte_extract_exprt`.

File: src/solvers/flattening/flatten_byte_operators.cpp [hunk 4]
```diff
@@ -517,9 +521,10 @@ exprt flatten_byte_update(
     }
     else
     {
-      throw
-        "flatten_byte_update can only do arrays of scalars right now, "
-        "but got "+subtype.id_string();
+      INVARIANT_WITH_DIAGNOSTICS(
+        false,
+        "flatten_byte_update should only be used on arrays of scalars",
+        subtype.id_string());
```
Comment (tautschnig) [inline src/solvers/flattening/flatten_byte_operators.cpp]: No, it can soundly be used on other types, it's just not yet implemented...

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d07bba74eb891c64080c8ac6e199b49f8aa993e9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85574428

**Comment by danpoe** [comment]
@tautschnig I think this is ready to be merged.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**